protected-api-access-protected-api/app.js

```diff
@@ -0,0 +1,119 @@
+/*
+ This application demonstrates how to issue a call to a protected API
+ using the On-behalf-of flow.  A request will be issued to the
+ Microsoft Graph /me endpoint on behalf of the authenticated user.
+*/
+
+// Microsoft Authentication Library (MSAL) for Node.Js
+const msal = require('@azure/msal-node')
+
+// Node.Js Express Framework
+const express = require('express')
+
+// Used to make the HTTP GET request to the Graph API
+const https = require('https')
+
+// Used to verify authentication tokens
+const jsonwebtoken = require('jsonwebtoken')
+const jwksRsa = require('jwks-rsa')
+
+// The port the express server will listen on
+const SERVER_PORT = 8080
+
+// 'Directory (tenant) ID' in Azure portal
+const tenant = ''
+
+const config = {
+  auth: {
+    
+    // 'Application (client) ID' of app registration in Azure portal - this value is a GUID
+    clientId: '',
+    
+    // Client secret 'Value' (not the ID) from 'Client secrets' in app registration in Azure portal
+    clientSecret: '',
+
+    // Full directory URL, in the form of https://login.microsoftonline.com/<tenant>
+    authority: `https://login.microsoftonline.com/${tenant}`
+  }
+}
+
+// Initialize MSAL configuration with above values
+const cca = new msal.ConfidentialClientApplication(config)
+
+const app = express()
+
+
+// The signing keys are located on the specified JWKS (JSON Web Key Set) endpoint.
+// Required for token validation.
+const jwksClient = jwksRsa({
+  jwksUri: `https://login.microsoftonline.com/${tenant}/discovery/v2.0/keys`
+})
+
+
+const validateJwt = (req, res, next) => {
+  const authHeader = req.headers.authorization
+
+  if (authHeader) {
+
+    // Extract the authorization token from the header
+    const token = authHeader.split(' ')[1]
+
+    // Each token has a target audience and issuer
+    const validationOptions = {
+      audience: config.auth.clientId,
+      issuer: config.auth.authority + '/v2.0'
+    }
+
+    // Attempt to verify the token. If validation fails, return an HTTP 401 error.
+    jsonwebtoken.verify(token, getSigningKeys, validationOptions, (err, payload) => {
+      if (err) {
+        console.log(err)
+        return res.sendStatus(401)
+      }
+      next()
+    })
+  }
+}
+
+const getSigningKeys = (header, callback) => {
+  jwksClient.getSigningKey(header.kid, function (err, key) {
+    const signingKey = key.publicKey
+    callback(null, signingKey)
+  })
+}
```

Comment:
Again, to try to "inline" as much as possible, can you instead do something more like.  (Note: this is just a freeform suggestion here, there might be syntax or other issues, just the "thought" is what I'm trying to convey here.)  If you can simplify it even more than what I have, the better.

```suggestion
    // Each token has a target audience and issuer
    const validationOptions = {
      audience: config.auth.clientId,
      issuer: config.auth.authority + '/v2.0'
    }
    
    // Look up the public key used to sign JWT tokens, as provided by the OIDC discovery endpoint
    const getSigningKey = (jwtHeader, callback) => {
        callback(null, jwksClient.getSigningKey(jwtHeader.kid))
      }
    }

    // Attempt to verify the token. If validation fails, return an HTTP 401 error.
    jsonwebtoken.verify(token, getSigningKeys, validationOptions, (err, payload) => {
      if (err) {
        console.log(err)
        return res.sendStatus(401)
      }
      next()
    })
  }
}
```

Reply:
👍